cai_causal_graph/time_series_causal_graph.py

```diff
@@ -0,0 +1,673 @@
+        In other words, it is a graph that has no edges whose destination is not time delta 0.
+        """
+        minimal_cg = TimeSeriesCausalGraph()
+
+        # first step
+        # go through all the nodes and add them to the minimal graph if they have time delta 0
+        for node in self.get_nodes():
+            if node.get_metadata()[self._meta_time_name] == 0:
+                # add if node is not already in the graph
+                if not minimal_cg.node_exists(node.identifier):
+                    minimal_cg.add_node(node=node)
+
+        # second step: add nodes that are connected to the nodes in the minimal graph
+        # go through all the nodes in the minimal graph and add the nodes that are entering them
+        for node in minimal_cg.get_nodes():
+            for edge in self.get_edges():
+                # could also check just edge.destination == node (avoiding the metadata check)
+                if edge.destination.identifier == node.identifier:
+                    # add the node if it is not already in the graph
+                    if not minimal_cg.node_exists(edge.source):
+                        minimal_cg.add_node(node=self.get_node(edge.source))
+                    minimal_cg.add_edge(edge=edge)
+
+        return minimal_cg
+
+    def is_minimal_graph(self) -> bool:
+        """Return True if the graph is minimal."""
+        return self == self.get_minimal_graph()
+
+    def get_summary_graph(self) -> CausalGraph:
+        """
+        Return a summary graph.
+
+        Collapse graph in time into a single node per variable (column name).
+        This can become cyclic and bi-directed as X-1 -> Y and Y-1 -> X would become X <-> Y.
+
+        There are several cases to consider. Assume the edge in consideration is called B.
+        - if the edge is not in the summary graph, add it
+        - if there's an edge A is in the summary graph, then:
+            - if A and B have the same direction, keep the direction
+            - if A and B have different directions, make it bi-directed
+            - if one of the two is already bi-directed, keep it bi-directed
+
+        :return: The summary graph as a CausalGraph object.
+        """
+        if self._summary_graph is None:
+            summary_graph = CausalGraph()
+            # now check as described above (assume egdes are already directed)
+            edges = self.get_edges()
+            for edge in edges:
+                # first we need to extract the variable names from the nodes as the summary graph
+                # will have the variable names as nodes
+                source_node = edge.source
+                destination_node = edge.destination
+                source_variable_name = source_node.get_metadata()[self._meta_variable_name]
+                destination_variable_name = destination_node.get_metadata()[self._meta_variable_name]
```

Comment:
again this would fail if it is not in the metadata, should that be caught explicitly?